[LUA] Create job_utils/paladin, migrate job functionality, and update JAs

Schaedelschaden opened the pull request:
<!-- Remove space and place 'x' mark between square [] brackets or click the checkbox after saving to affirm the following points: -->
<!-- (it should look like this: - [x] I have ...) -->
**_I affirm:_**
- [x] I have paid attention to this example and will edit again if need be to not break the formatting, or I will be ignored
- [x] I have read and understood the [Contributing Guide](https://github.com/LandSandBoat/server/blob/base/CONTRIBUTING.md)
- [x] I've _**tested my code and the things my code has changed**_ since the last commit in the PR, and will test after any later commits

## What does this pull request do?

<!-- Describe what your PR does here. If it closes an existing issue, you can mention: "Closes #1234" and GitHub will link this PR to that issue. -->

Restores accidentally closed #2734.

**Job Abilities**
- Create job_utils/paladin.lua and populates necessary ability check and ability use functions.
- Moves functionality from all Paladin globals/abilities files to globals/job_utils/paladin.lua.
- Updates Divine Emblem application in magic.lua and damage_spell.lua.
- Sets up "Enhances 'Divine Emblem' effect" mod to adjust Divine Emblem enmity bonus.
- Sets up Intervene to check for equipped shield, do damage, and inflict Intervene status effect on target.
- Sets up Majesty to apply Cure Potency II and Cure Recast mods, and make Cure and Protect spells AoE.
- Sets up Sepulcher to be usable against Undead enemies only.
- NOTE: Additional changes will be added to this list as they occur.

**Effects**
- Converts Fealty to check resistance inside the globals/magic.lua applyResistanceEffect function.
- Updates Fealty status effect to be dispelable.
- Sets up the Intervene status effect and debuffs.
- Sets up the Majesty status effect and cure potency/white magic recast bonuses.
- Sets up the Palisade status effect and shield block bonus.
- Converts Rampart to SDT and adds the Rampart "Enhances 'Iron Will' effect" fast cast bonus.
- Converts Reprisal effect bonuses to the core battleutils.cpp GetBlockRate function.
- Reduces the enmity bonus from Sentinel by 50 if Paladin is not the player's main job.
- Sets up the Sepulcher status effect and debuffs.

**Spells**
- Adds Amnesia effect to Holy II when caster has Divine Emblem.

**Augments**
- Sets up "Enhances 'Iron Will' effect" mod and gear augment (Valor/Caballarius Coronet).
- Sets up "Enhances 'Fealty' effect" mod and gear augment (Valor/Caballarius Surcoat).
- Sets up "Enhances 'Chivalry' effect" mod and gear augment (Valor/Caballarius Gauntlets).
- Sets up "Enhances 'Invincible' effect" mod and gear augment (Valor/Caballarius Breeches).
- Sets up "Enhances 'Guardian' effect" mod and gear augment (Valor/Caballarius Leggings).

**Core/LUA Functions**
- Adds lua_spell.cpp getSpellFamily function for pulling a spell's SPELLFAMILY.
- Adds battleutils.cpp GetSpellAoEType "Caster in party" check for Majesty and Accession.
- Adds spell.cpp getAoEMessage single target -> AoE "recovers HP" message override.
- Adds magic_state.cpp ApplyEnmity Divine Emblem enmity handling and status removal.

## References

All job abilities - https://www.bg-wiki.com/ffxi/Paladin
Shield Block Rate - https://www.bg-wiki.com/ffxi/Shield_Skill
Reprisal - https://www.bg-wiki.com/ffxi/Reprisal

## Steps to test these changes

1. Upload changes to your server.
2. Load the server.
3. Use the `!changejob PLD 99` command.
4. Use all JA's to confirm they do not throw errors.
5. Test all JA's and compare their behavior to BGWiki JA descriptions.

Schaedelschaden committed:
[LUA] Create job_utils/paladin, migrate job functionality, and update JAs

Schaedelschaden commented:
Crap, I apologize. I thought this was on my repository.

WinterSolstice8 commented on scripts/globals/job_utils/paladin.lua:
```diff
@@ -0,0 +1,223 @@
+end
+
+xi.job_utils.paladin.useHolyCircle = function(player, target, ability)
+    -- TODO: Create xi.mod.UNDEAD_SDT mod and core handling?
```
probably don't need undead SDT, just need to check undead family vs holy circle effect power

WinterSolstice8 replied:
though on second thought a mod is much faster to lookup in theory... out of scope of this PR

Schaedelschaden replied:
My main thought with making them mods was the future addition of the other Circle effect SDT’s and killer effect equivalent SDT’s. It would need to be adjustments in the core handling for damage that are pulled from the player (or jug/charmed pet) and would increase damage and reduce damage taken against that family.

Thinking about it further, calling them SDT’s may be a misnomer. They do affect a specific damage category though.

I haven’t gone back through and investigated how to accomplish the set up yet, definitely out of the scope of this PR. I can set up an issue with the TODO and my findings when the PR is merged.

TeoTwawki replied:
> Thinking about it further, calling them SDT’s may be a misnomer. They do affect a specific damage category though.

the language is def a problem, because sdt means a very different thing in our core.

TeoTwawki replied:
can we change the comment here to just reflect the unimplemented part with info from wiki without a suggested modifier name? like
```lua
-- 15% damage bonus against 15% damage resistance from undead, 5% when subbed
```
It's really a kind of damage _given_ mod combined with a not-specific damage taken

more a `dmg_vs_eco` in both directions really

Schaedelschaden replied:
Addressed in #2847.

WinterSolstice8 reviewed (approved):
LGTM

WinterSolstice8 commented:
Is this still meant to be a draft?

Schaedelschaden commented:
I was planning to add additional changes to the PR for further functionality. The previous PR got messed up when I tried to update the HEAD after a commit gave me an error.

Schaedelschaden commented:
Apologies, my wife and I had our first child two weeks ago and have been trying to get back into a regular pattern again so the PR has been taking a bit to get put together.

TeoTwawki commented:
> Apologies, my wife and I had our first child two weeks ago and have been trying to get back into a regular pattern again so the PR has been taking a bit to get put together.

Take your time, and congratulations

Schaedelschaden commented:
I keep getting a sanity check failure for modifier.h on this commit.

`HEAD is now at 1b7df52e4b Merge a8d97281702734d860a8b496ffb0838bab86830b into f62423eb1caa5f599e8af9ef13fa5d15fac5426d`

I’ve tried switching back to my LSB base branch, updated to the most recent commit, switched back to my pld_job_utils_1 branch, rebased and merged to base, receive the message that the HEAD is up to date, and still receive the HEAD error when I make the commit.

I’m understanding that it’s wanting me to merge my most recent commit to LSB base’s most recent commit but I’m not sure why it’s telling me everything is up to date and still throwing this error in the sanity check. Am I missing forcing the pointer to update?

WinterSolstice8 commented:
> and still receive the HEAD error when I make the commit.
what's the actual output?

Schaedelschaden committed:
Create new PLD mods

Schaedelschaden commented:
I think I had rebased incorrectly. Updating my base to the current HEAD, then switching back to pld_job_utils_1 and rebasing to origin seems to have satisfied it. Thank you Winter!

Schaedelschaden committed:
Update Fealty functionality

Schaedelschaden committed:
Update Intervene & Majesty effects and abilities.sql

Schaedelschaden commented:
Hmm, it's failing the sanity check with the HEAD error again.

My git bash:
![mintty_tSWw4IyvL1](https://user-images.githubusercontent.com/16365541/192422531-97de4c8d-b0f4-468f-83d0-93b4294bbf20.png)

The sanity check:
![firefox_sEpFkrO71j](https://user-images.githubusercontent.com/16365541/192422556-2d350cf9-cf57-4652-a2bb-b18f83dc99b0.png)

WinterSolstice8 commented:
"HEAD" isn't an error. what you're seeing is that it's telling you how it expects the formatting. Looks like spaces at a glance.

Schaedelschaden commented:
Ooooh, is it a formatting issue? It's expecting the spacing to be the way it's showing?

WinterSolstice8 commented:
> Ooooh, is it a formatting issue? It's expecting the spacing to be the way it's showing?

yup thats right. the "-" line is what your line is, and the "+" line is what it says it should be

Schaedelschaden commented:
Well crap, I feel like...well, not something nice. /headdesk, let's give this another try.

Schaedelschaden committed:
Add core handling for Majesty AoE and message adjustment

WinterSolstice8 commented:
It looks like you're going on a full blown job audit with these now, very nice.

Schaedelschaden commented:
Yessir, I figured if I was going to go through and update the things that weren't working, I might as well hit as much as possible!

Schaedelschaden committed:
Update Palisade & Reprisal functionality and shield block rate calcs

Schaedelschaden committed:
Update Rampart, Sentinel, and Sepulcher effects

Schaedelschaden commented:
Well, that's a new one.

![firefox_qlwmaB6k5x](https://user-images.githubusercontent.com/16365541/192481999-9535a285-e487-42fb-b132-bb55d1db30e5.png)

Not sure how to reduce the complexity of the function, all I did was add another status effect check. Should the various status checks be combined somehow?

Schaedelschaden committed:
Update Divine Emblem functionality

Schaedelschaden committed:
Update PLD specific augments

WinterSolstice8 commented on src/map/utils/battleutils.cpp:
```diff
@@ -5595,11 +5614,53 @@ namespace battleutils
-            if (PCaster->StatusEffectContainer->HasStatusEffect(EFFECT_ACCESSION) ||
-                (PCaster->objtype == TYPE_PC && charutils::hasTrait((CCharEntity*)PCaster, TRAIT_DIVINE_VEIL) && PSpell->isNa() &&
-                 (PCaster->StatusEffectContainer->HasStatusEffect(EFFECT_DIVINE_SEAL) || PCaster->getMod(Mod::AOE_NA) == 1)))
+            if (casterInParty == true &&
```
Accession explicitly works outside your party, same with Divine Veil. Accession is pretty famous for being able to AoE cure _all_ NPC allies for easy points in Campaign as well

bg wiki confirms this as well. I think it's just Majesty that doesn't work like that.

WinterSolstice8 replied:
On second thought, Majesty is too new and unknown to me to just make a claim that it doesn't work in other parties.

I'll go test it.

Schaedelschaden replied:
Confirmed with retail testing that both Accession and Majesty will activate as AoE heals on other parties.

Schaedelschaden committed:
Remove caster in party requirement from Majesty and Accession

WinterSolstice8 reviewed (approved):
Nothing stands out to me as wrong, and the things that did were recent changes that check out.

The unknowns are documented with TODOs with what appears to be reasonably guessed, such as Divine Emblem granting +100 MACC for divine spells which is not all that unreasonable given elemental seal is likely far greater than that.
